### PR TITLE
Search: add support for adding wpcom plan to the shopping cart

### PR DESCRIPTION
**Pull request by AnnaMag**
#### Changes proposed in this Pull Request

Changes:
* enable adding Search for WP.com and AT sites

#### Testing instructions

_FULL flow:_
- go to `jetpack/connect/jepack_search`,  `/wpcom_search`, `/jepack_search/monthly`,  `/wpcom_search/monthly`, 
- start typing a WP.com site (without the product) and choose it from the selection
- proceed to checkout and purchase the product
- verify that a Thank you screen is present after the purchase
- verify that track events fire: `calypso_cart_product_add` with correct product slug prop, verify network request with the correct product info

_Testing adding a correct product:_
- /checkout/:site/jetpack_search|wpcom_search|jetpack_search_monthly|wpcom_search_monthly irrespective of the product slug, show correct product type in the cart (wpcom_search for AT and wpcom, and jetpack_search for JP)

### CODE REVIEW

**Comment by DavidRothstein** [comment]
>go to `jetpack/connect/jetpack_search` or `/wpcom_search`

This isn't a real review for now, but just wanted to point out that for the above, if I go to `jetpack/connect/jetpack_search` and select a WPCOM site, it puts the Jetpack version of the search product in the shopping cart (rather than the WPCOM version).  That leads to an error when trying to pay, because the product isn't compatible with the site.

For the opposite case (if I go to `jetpack/connect/wpcom_search` and select a Jetpack site) it redirects me to the Jetpack plans page for some reason.

Seems like the code might need to be made smart enough to always use the `jetpack_search` product for Jetpack sites and `wpcom_search` product for WPCOM sites, regardless of what's in the URL?

**Comment by AnnaMag** [comment]
Thanks, @DavidRothstein! Addressed.

**Comment by gibrown** [comment]
I was able to purchase on an AT site using this. I think the main ones are:
- went to `jetpack/connect/jetpack_search` and found my site. When I selected it and clicked on "Get Search" i got redirected back to the `jetpack/connect/jetpack_search` url. Search did get added to my shopping cart though
- Currently the my-plans page also is not showing the Search plan after I purchase it. I had thought I saw that working. Different PR?

I also tried going directly to `checkout/.../jetpack_search` and that also gave me a redirect

I also tried going to a wpcom_search url and got redirected (which is probably fine).

**Comment by AnnaMag** [comment]
@gibrown, thanks for testing! 

The redirects should be fixed now. 

From my testing, the product is added to /my-plan correctly for Jetpack sites (jetpack_search slug), but missing for the wpcom sites. Could you confirm?

**Comment by gibrown** [comment]
Yep, redirects fixed now. Thanks.

I tried to buy from the `jetpack/connect/jetpack_search` which then brought me to `checkout/gibrownatomictest.wpcomstaging.com/jetpack_search` (correctly). But after purchasing I still couldn't see a Search plan on the my-plan page.

I think the slug will always end up being wpcom_search on a wp.com site though right?

Should we break issues like this out into separate PRs/issues? Maybe specifically have an issue for each of the screens we know have some problems?

**Comment by AnnaMag** [comment]
@gibrown, thanks for reviewing. It is ready for final round testing.

> I think the slug will always end up being wpcom_search on a wp.com site though right?

We made determining the product to be slug independent, based on the type of site, not the slug itself. So wpcom_ and jetpack_ are used interchangeably.

**Comment by AnnaMag** [comment]
Added feature flag for development and stage envs.

**Comment by gibrown** [comment]
Minor paths that don't seem to work:
- /jetpack/connect/jetpack_search_monthly
- /jetpack/connect/wpcom_search_monthly

File: client/jetpack-connect/jetpack-connect-notices.jsx [hunk 1]
```diff
@@ -107,14 +105,6 @@ export class JetpackConnectNotices extends Component {
 				noticeValues.text = translate( 'Good news! WordPress.com sites already have Jetpack.' );
 				return noticeValues;
 
-			case IS_DOT_COM_GET_SEARCH:
```
Comment (gibrown) [inline client/jetpack-connect/jetpack-connect-notices.jsx:110]: Shouldn't we be keeping this until it is available in production?

File: client/jetpack-connect/jetpack-connection.jsx [hunk 2]
```diff
@@ -197,12 +198,26 @@ const jetpackConnection = ( WrappedComponent ) => {
...
+				return IS_DOT_COM;
+			}
+
 			if ( this.isError( 'site_blacklisted' ) ) {
 				return SITE_BLACKLISTED;
 			}
 
-			if ( this.checkProperty( 'userOwnsSite' ) ) {
-				return ALREADY_OWNED;
+			if ( this.checkProperty( 'isWordPressDotCom' ) ) {
```
Comment (gibrown) [inline client/jetpack-connect/jetpack-connection.jsx:214]: duplicated code from above?

File: client/my-sites/checkout/checkout/index.jsx [hunk 3]
```diff
@@ -306,13 +306,25 @@ export class Checkout extends React.Component {
...
 		) {
 			cartItem = jetpackProductItem( this.props.product );
 		}
 
+		if (
+			startsWith( this.props.product, 'jetpack_search' ) ||
+			startsWith( this.props.product, 'wpcom_search' )
+		) {
+			if ( isJetpackNotAtomic ) {
+				cartItem = jetpackProductItem( 'jetpack_search' );
```
Comment (gibrown) [inline client/my-sites/checkout/checkout/index.jsx:320]: I think this may break purchasing the monthly plan for normal jetpack sites correct?

**Review by gibrown — CHANGES_REQUESTED** [review]
a couple of questions

File: client/jetpack-connect/jetpack-connection.jsx (hunk 2 above)
Comment (AnnaMag) [reply client/jetpack-connect/jetpack-connection.jsx:214]: yes!

File: client/jetpack-connect/jetpack-connect-notices.jsx (hunk 1 above)
Comment (AnnaMag) [reply client/jetpack-connect/jetpack-connect-notices.jsx:110]: makes sense

File: client/my-sites/checkout/checkout/index.jsx (hunk 3 above)
Comment (AnnaMag) [reply client/my-sites/checkout/checkout/index.jsx]: addressed

**Comment by AnnaMag** [comment]
Feedback has been addressed and the monthly slug should be working now. 
I updated the testing instructions accordingly.

**Review by gibrown — APPROVED** [review]
Tested out a number of purchase options both for the production build and dev builds. I think everything is working well enough that we should merge this in and then iterate on the remaining pieces/bugs. Existing purchases of Search should all still work.

**Review by wp-desktop — DISMISSED** [review]
WordPress Desktop CI Failure for job "wp-desktop-source".

@AnnaMag please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/795247). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso master.

**Comment by nsakaimbo** [comment]
@AnnaMag Looks like there's some build errors in the job linked by that wp-desktop review:

```
$ check-npm-client && node ${NODE_ARGS:---max_old_space_size=8192} ./node_modules/webpack/bin/webpack.js --config client/webpack.config.js --display errors-only && yarn run build-languages-if-enabled
 
ERROR in ./jetpack-connect/jetpack-connection.jsx
 
Module build failed (from ../node_modules/babel-loader/lib/index.js):
SyntaxError: /home/circleci/wp-calypso/client/jetpack-connect/jetpack-connection.jsx: Unexpected token (8:1)
   6 | import React, { Component } from 'react';
   7 | import { connect } from 'react-redux';
>  8 | <<<<<<< HEAD
     |  ^
   9 | import { flowRight, get, includes, startsWith, omit } from 'lodash';
  10 | =======
  11 | import { flowRight, get, includes, startsWith } from 'lodash';
    at Object._raise (/home/circleci/wp-calypso/node_modules/@babel/parser/lib/index.js:742:17)
    at Object.raiseWithData (/home/circleci/wp-calypso/node_modules/@babel/parser/lib/index.js:735:17)
    at Object.raise (/home/circleci/wp-calypso/node_modules/@babel/parser/
```

**Comment by AnnaMag** [comment]
@nsakaimbo thanks for the ping! Yes, I pushed an unfinished rebase and was correcting it. All under control :)

**Pull request merged**